rs/execution_environment/src/canister_manager.rs

```diff
@@ -1914,9 +1914,10 @@ impl CanisterManager {
 
         // Delete old snapshot identified by `replace_snapshot` ID.
         if let Some(replace_snapshot) = replace_snapshot {
+            let old_snapshot = state.canister_snapshots.remove(replace_snapshot);
             // Already confirmed that `replace_snapshot` exists.
-            let is_removed = state.canister_snapshots.remove(replace_snapshot);
-            debug_assert!(is_removed.is_some());
+            debug_assert!(old_snapshot.is_some());
+            canister.system_state.snapshots_memory_usage -= old_snapshot.unwrap().size();
```

Comment:
We should make all subtraction a saturating subtraction. 

It would be a bug if it ever makes a difference, but if it happens we definitely shouldn't wrap around.

Reply:
One other thing we do in similar cases (e.g. throughout `CanisterQueues`) is to have a `debug_assert_eq!(self.computed_<field>(), self.field)`. It gets compiled out in production code, but makes any unit test that touches this code also a unit test for the respective field.

Reply:
FWIW, I don't see the `debug_assert_eq!()` I suggested above _after_ every mutation to `snapshots_memory_usage`.

I don't know if that is intentional (because you added a different `debug_assert_eq!()` just before mutating `snapshots_memory_usage`), but my suggested approach would be more general (including if/when we later allow for multiple snapshots per canister).

Not insisting it's done my way, just clarifying whether this was intentional.